### ISSUE TITLE
Math.imul on stdlib.js was not properly evaluated in browserless Javascript engine

### ISSUE THREAD

**Issue by jmineraud** [ocsigen/js_of_ocaml#255]
Hi,

I have noticed when executing Javascript code produced by JSOO with the Duktape ECMAScript engine that Math.imul was not declared properly. By adding the declaration from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/imul to my script, I was able to fix the issue.

However, it seems that stdlib.js is already trying to do the same job, so it might be a bug

Best Regards

**Comment by hhugo** [comment]
related to https://github.com/ocsigen/js_of_ocaml/commit/ba043eaf

**Comment by vouillon** [comment]
Hi,
Indeed, we are supposed to use a polyfill implementation if Math.imul is not defined.
So, what do you observe? Do you get an error, or do you get wrong results?
Do you have a short piece of code which reproduces the bug? Or just two integer which are incorrectly multiplied.

**Comment by hhugo** [comment]
@jmineraud , what do you mean by `not declared properly`

**Comment by jmineraud** [comment]
Hi,
I don't have a small piece of code that would reproduce the error, but when I execute a javascript produced by JSOO (I put the lowest level of optimization and pretty print to find where the error was), which only does a XMLHttpRequest, in a javascript engine such as Duktape, I need to define the XMLHttpRequest function (which is fine) and the global reference to location (which is less needed). However, one more function (Math.imul) was required an by the JSOO output but could not be found. However, I checked the JSOO output and found the function:

```javascript
/*<<stdlib.js 221 0>>*/ if(!Math.imul)
     Math.imul=
     function(x_a_,y_b_)
      { /*<<stdlib.js 223 21>>*/ return ((x_a_>>16)*y_b_<<16)+
              (x_a_&65535)*
              y_b_|
              0 /*<<stdlib.js 223 4>>*/ };
```
I was able to fix the issue by declaring myself the Math.imul function using the code from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/imul
After this fix, running the script in the Duktape engine worked. I concluded that the definition of Math.imul by JSOO may not work in these browserless environments.

The ocaml code I used to generate the JS code is the following:
```ocaml
open Js
open Lwt
open XmlHttpRequest

let print msg =
  Js.Unsafe.fun_call (Js.Unsafe.variable "print") [|Js.Unsafe.inject (Js.string msg)|]

let debug msg = 
	Firebug.console##log(Js.string msg)

let (>>=) = Lwt.bind

exception Wrong_request_type of string

let xml_http_request request_type url data async =
	let _ = 
		match request_type with
		| "GET" | "POST" | "DELETE" | "PUT" -> ()
		| _ -> raise (Wrong_request_type request_type)   
	in
	let (res, w) = Lwt.task () in
	let req = create () in
	let headers s =
		Opt.case
		(req##getResponseHeader (Js.bytestring s))
		(fun () -> None)
		(fun v -> Some (Js.to_string v))
	in
	req##_open (Js.string request_type, Js.string url, Js.bool async);
	req##setRequestHeader (Js.string "Content-type", Js.string ("text/plain;charset=UTF-8"));
	req##onreadystatechange <- Js.wrap_callback
		(fun _ ->
			(match req##readyState with
			| DONE ->
				Lwt.wakeup w
					{url = url;
					code = req##status;
					content = Js.to_string req##responseText;
					content_xml =
						(fun () ->
							match Js.Opt.to_option (req##responseXML) with
							| None -> None
							| Some doc ->
								if (Js.some doc##documentElement) == Js.null
								then None
								else Some doc);
					headers = headers
					}
			| _ -> ()));
	(match data with
	| None -> req##send (Js.null)
	| Some d -> req##send(Js.some(Js.string d))
 	);
	Lwt.on_cancel res (fun () -> req##abort ());
	res

let http_method request_type url data async =
	xml_http_request request_type url data async >>= fun response ->
	let code_ = response.XmlHttpRequest.code in
	(*let type_ = response.XmlHttpRequest.headers "Content-type" in*)
	let msg = response.XmlHttpRequest.content in
	if (code_ = 0 || code_ = 200)
	then Lwt.return (msg)
	else fst (Lwt.wait ())

let http_get url async =
	http_method "GET" url None async
let _ =
	http_get "http://127.0.0.1:8080/" true >|= (fun msg ->		
		print ("TEST XMLHttpRequest JSOO: /GET http://127.0.0.1:8080/ -> " ^ msg)
	)
```

**Comment by vouillon** [comment]
Our function does not give the right result when the second argument is not an integer. Maybe that's the issue here. Though I don't really understand why it would be called with non-integer arguments.

**Comment by vouillon** [comment]
I have fixed our polyfill in commit 237a9de61fa663f2ec17638eeb06ffda5b117518.
@jmineraud, it would be great if you could check whether this fixes your issue.

**Comment by jmineraud** [comment]
Hi, I have checked with the new commit and it passes the test. I also had the update the javascript engine to the latest dev because I saw a recent commit that was fixing a similar issue. I think the problem was coming from there.
